python/lsst/daf/butler/tests/_testRepo.py

```diff
@@ -388,32 +349,71 @@ def addDataIdValue(butler: Butler, dimension: str, value: Any, **related: Any):
     except KeyError as e:
         raise ValueError from e
     # Bad keys ignored by registry code
-    extraKeys = related.keys() - (fullDimension.required | fullDimension.implied)
+    extraKeys = related.keys() - fullDimension.graph.dimensions.names
     if extraKeys:
         raise ValueError(
-            f"Unexpected keywords {extraKeys} not found "
-            f"in {fullDimension.required | fullDimension.implied}"
+            f"Unexpected keywords {extraKeys} not found " f"in {fullDimension.graph.dimensions.names}"
         )
 
-    if fullDimension.viewOf:
-        # Nothing to do; this dimension's records aren't actually saved in the
-        # database directly anyway (and caller shouldn't have to know this).
-        return
-
-    # Define secondary keys (e.g., detector name given detector id)
-    expandedValue = _fillAllKeys(fullDimension, value)
-    expandedValue.update(**related)
-    completeValue = _fillRelationships(fullDimension, expandedValue, butler.registry)
-
-    dimensionRecord = fullDimension.RecordClass(**completeValue)
-    try:
-        butler.registry.syncDimensionData(dimension, dimensionRecord)
-    except sqlalchemy.exc.IntegrityError as e:
-        raise RuntimeError(
-            "Could not create data ID value. Automatic relationship generation "
-            "may have failed; try adding keywords to assign a specific instrument, "
-            "physical_filter, etc. based on the nested exception message."
-        ) from e
+    # Assemble a dictionary data ID holding the given primary dimension value
+    # and all of the related ones.
+    data_id: dict[str, Union[int, str]] = {dimension: value}
+    data_id.update(related)
+
+    # Compute the set of all dimensions that these recursively depend on.
+    all_dimensions = butler.registry.dimensions.extract(data_id.keys())
+
+    # Create dicts that will become DimensionRecords for all of these data IDs.
+    # This iteration is guaranteed to be in topological order, so we can count
+    # on new data ID values being invented before they are needed.
+    record_dicts_by_dimension: dict[Dimension, dict[str, Any]] = {}
+    for dimension_obj in all_dimensions:
```

Comment:
I'm finding this loop very hard to follow because of the lack of distinction between the dimension we are actually trying to create (`data_id`) and the dimensions that are incidental to that and probably already exist (`dimension_obj`, `dimension_value`, etc.). I especially don't understand why you are carefully constructing dimension records for existing dimensions.

Would it be possible to factor out some of these blocks into their own functions (making the code a bit more self-documenting in the process)? It looks like the body of `if dimension_value is None` and the code below the view check are each trying to do one thing. This should reduce the number of variables the human needs to keep track of at a time.

Reply:
In a system where any missing dependencies are created as needed, I don't think there is a distinction between the original dimension identified and any others anymore; it's unfortunate that the interface continues to suggest that there is, but I don't want to go through the deprecation-cycle effort of changing that now, either.

>  especially don't understand why you are carefully constructing dimension records for existing dimensions.

I don't think I am; I am fetching any records that exist that are compatible with the given data IDs.  Calling `syncDimensionData` on those later should be a no-op (it's an expensive enough one that I would try to avoid it in non-test code, but here I figured that was better than having another local variable to track which records are new and which are existing).

> Would it be possible to factor out some of these blocks into their own functions

:+1: